Add Landsat 1-5 (MSS and TM) and 7 (ETM) Item creation

pjhartzell opened the pull request:
Adds functions (all in `usgs_stac.py`) to create new STAC Items from one or more USGS `stac_version: 1.0.0` Items. 
- Uses same input as current landsat-8 method (a single `*_MTL.xml` file href)
- Tested on an MSS (Landsat 1-5, Collection 2, Level-1) `xml`
- Tested on a TM (Landsat 4-5, Collection 2, Level-2) `xml`
- Note the tests currently create the example MSS and TM STAC items

gadomski commented on examples/usgs-stac-convert/mss/LM01_L1GS_001010_19720908_02_T2.json:
```diff
@@ -0,0 +1,318 @@
+    },
+    {
+      "rel": "self",
+      "href": "/Users/pjh/dev/landsat/examples/usgs-stac-convert/mss/LM01_L1GS_001010_19720908_02_T2.json",
```
We should save examples with `include_self_href=False` to avoid local file paths.

pjhartzell replied:
Agreed. Updated.

gadomski commented on examples/usgs-stac-convert/mss/LM01_L1GS_001010_19720908_02_T2.json:
```diff
@@ -0,0 +1,318 @@
+  ],
+  "assets": {
+    "thumbnail": {
+      "href": "tests/data-files/usgs-stac/mss/LM01_L1GS_001010_19720908_20200909_02_T2_thumb_small.jpeg",
```
Try using `make_asset_hrefs_relative()` to get valid relative paths, e.g. https://github.com/stactools-packages/modis/blob/main/scripts/create_expected.py#L35.

pjhartzell replied:
Hmm. Setting `item.make_asset_hrefs_relative()` does not change the output. Are you thinking that it should show something along the lines of `../../tests/data-files/...`? If I use `item.make_asset_hrefs_absolute()`, then the path is full, e.g., `/Users/pjh/dev/...`.

gadomski replied:
Yeah, `make_asset_hrefs_relative` requires (someone counterintuitively) an absolute path to work. I pushed up a fix here: https://github.com/pjhartzell/landsat/commit/756703dbe357fe43d9d07853c63b9bd36124be1f.

gadomski commented on tests/test_usgs_stac.py:
```diff
@@ -0,0 +1,40 @@
+class UsgsStacTest(unittest.TestCase):
+
+    def test_tm(self):
+        mtl_xml_href = "tests/data-files/usgs-stac/tm/LT04_L2SP_002026_19830110_20200918_02_T1_MTL.xml"  # noqa
```
To guard against relative filepath issues if tests are run from the non-root directory, use https://github.com/stac-utils/stactools/blob/main/src/stactools/testing/test_data.py#L22.

pjhartzell replied:
Ah. I see and understand. Will do.

gadomski commented on tests/test_usgs_stac.py:
```diff
@@ -0,0 +1,40 @@
+        item = convert_usgs_stac(mtl_xml_href)
+        item.validate()
+
+        destination = "examples/usgs-stac-convert/tm"
```
I don't think we should create the `examples` during unit test execution, because that would update e.g. the `created` attribute every test run. A `scripts/create_examples` would be more appropriate. Here in the unit tests, use a `TemporaryDirectory`.

pjhartzell replied:
Yep. No longer using tests to create the examples.

gadomski commented on src/stactools/landsat/stac.py:
```diff
@@ -1,102 +1,281 @@
+    sensor: Sensor,
+    product_id: str,
+    read_href_modifier: Optional[ReadHrefModifier] = None
+) -> Union[Dict[str, Any], None]:
```
```suggestion
) -> Optional[Dict[str, Any]]:
```

pjhartzell replied:
Done.

gadomski commented on src/stactools/landsat/stac.py:
```diff
@@ -1,102 +1,281 @@
+    Returns:
+        Union[Dict[str, Any], None]: Either a GeoJSON geometry or None
+    """
+    # Check Planetary Computer storage first
```
How is this defaulting to Planetary Computer? `base_href` isn't optional, so it always has to be passed into the function.

pjhartzell replied:
Yep. Corrected.

gadomski commented on src/stactools/landsat/stac.py:
```diff
@@ -1,102 +1,281 @@
+            collection = "landsat-c2l2-sr"
+            product_id = f"{product_id}_SR"
+
+        catalog = Client.open("https://landsatlook.usgs.gov/stac-server")
```
Should probably be a constant.

pjhartzell replied:
Done.

gadomski commented on src/stactools/landsat/stac.py:
```diff
@@ -1,102 +1,281 @@
+        # -- USGS link
+        instrument = "-".join(i for i in SENSORS[sensor.value]["instruments"])
+        usgs_item_page = (
+            f"https://landsatlook.usgs.gov/stac-browser/collection02/level-"
```
Should probably be a constant.

pjhartzell replied:
Done.

gadomski commented on src/stactools/landsat/stac.py:
```diff
@@ -1,102 +1,281 @@
+        # component needs to have the processing time in it (replace item_id
+        # with product_id to fix. Needs to be resolved via a GitHub issue.
+        usgs_item_page = (
+            f"https://landsatlook.usgs.gov/stac-browser/collection02/level-2/standard/oli-tirs"
```
Should probably be a constant (or partially a constant).

pjhartzell replied:
Done.

pjhartzell committed:
adds STAC Item creation for MSS, TM, and ETM data

pjhartzell committed:
edit command documentation

pjhartzell committed:
add documentation

pjhartzell committed:
make oli-tirs fragment directory name compatable (oli_tirs)

pjhartzell committed:
add landsat-c2-l1 collection

pjhartzell committed:
change asset keys to common names, add create_collection command

pjhartzell committed:
finish collections, update readme and tests

pjhartzell committed:
make docker cibuild work

pjhartzell committed:
revert legacy_l8 to make existing tests untouched.

pjhartzell committed:
add some tests, update temporal extents, fix default values

pjhartzell committed:
fix typing error

gadomski commented on examples/landsat-c2-l1/LM01_L1GS_001010_19720908_02_T2/LM01_L1GS_001010_19720908_02_T2.json:
```diff
@@ -0,0 +1,301 @@
+    "sci:doi": "10.5066/P9AF14YV",
+    "landsat:cloud_cover_land": 46.0,
+    "landsat:wrs_type": "1",
+    "landsat:wrs_path": "001",
+    "landsat:wrs_row": "010",
```
Stray observation -- should these be `numeric`? I'm assuming they're provided as strings by USGS?

pjhartzell replied:
Good question. I see that both the USGS and Microsoft treat them as strings in their STAC Items. I'll be a lemming.

gadomski commented on examples/landsat-c2-l1/LM01_L1GS_001010_19720908_02_T2/LM01_L1GS_001010_19720908_02_T2.json:
```diff
@@ -0,0 +1,301 @@
+  ],
+  "assets": {
+    "thumbnail": {
+      "href": "tests/data-files/mss/LM01_L1GS_001010_19720908_20200909_02_T2_thumb_small.jpeg",
```
Still need to make these asset hrefs relative to the file (i.e. make them absolute then use `make_asset_hrefs_relative`).

pjhartzell replied:
OK, this is sinking in. I am now making the asset hrefs absolute during asset creation time, which allows the `make_all_asset_hrefs_relative()` call to do it's thing when creating collections.

gadomski commented on src/stactools/landsat/commands.py:
```diff
@@ -1,69 +1,169 @@
-    """Creates a command group for working
-    with Landsat metadata from USGS' Collection 2
+def create_landsat_command(cli: Group) -> Command:
+    """Creates a command group for commands working with the USGS' Landsat
```
```suggestion
    """Creates a command group for commands working with the USGS's Landsat
```

pjhartzell replied:
corrected

gadomski commented on src/stactools/landsat/commands.py:
```diff
@@ -1,69 +1,169 @@
-    def create_item_cmd(level: str, mtl: str, output: str,
-                        usgs_geometry: bool):
-        """Creates a STAC Item for a Landsat 8 C2 Level-2 scene's products.
+    @click.option("-l",
+                  "--legacy_l8",
+                  default=True,
+                  show_default=True,
+                  help="Create deprecated Landsat 8 STAC Item")
+    def create_item_cmd(level: str, mtl: str, output: str, usgs_geometry: bool,
+                        legacy_l8) -> None:
```
We should work on this wording. If its deprecated, it shouldn't be on by default.

pjhartzell replied:
Yeah. I didn't want any new defaults to change the current behavior at the moment. Thus far, if you use things as before, it will produce the same results. Trying not to break anyone's current workflow.  

I'll leave the `create_stac_item` default for using the legacy_l8 method as `True`, but change the command option to a flag, which makes the command option default to `False` for legacy_l8.

I also replaced the "deprecated" terminology with "legacy" to keep things consistent.

pjhartzell replied:
"deprecated" has been purged. Words have meaning.

gadomski commented on src/stactools/landsat/commands.py:
```diff
@@ -1,69 +1,169 @@
                   required=True,
                   help="HREF of directory in which to write the item.")
-    @click.option("--usgs_geometry",
+    @click.option("-u",
+                  "--usgs_geometry",
```
We should define this as a flag: https://click.palletsprojects.com/en/8.0.x/options/#boolean-flags.

pjhartzell replied:
Sounds good. Done.

gadomski commented on src/stactools/landsat/commands.py:
```diff
@@ -1,69 +1,169 @@
-    def create_item_cmd(level: str, mtl: str, output: str,
-                        usgs_geometry: bool):
-        """Creates a STAC Item for a Landsat 8 C2 Level-2 scene's products.
+    @click.option("-l",
+                  "--legacy_l8",
+                  default=True,
+                  show_default=True,
+                  help="Create deprecated Landsat 8 STAC Item")
```
Same as above, this could be defined as a flag: https://click.palletsprojects.com/en/8.0.x/options/#boolean-flags.

pjhartzell replied:
Sounds good. Done.

pjhartzell committed:
sundry review edits

- switch legacy_l8 command option to a flag, which changes the command default for using the legacy l8 method to to False
- remove "deprecated" terminology
- grammar edit

pjhartzell committed:
make asset hrefs absolute when creating assets

- allows the hrefs to then be made relative

pjhartzell committed:
remove relative asset paths from item command, import formatting

pjhartzell committed:
remove "Science Product" and "Data Product" from Ttem descriptions

pjhartzell committed:
replace alternate link with via links to usgs stac server

pjhartzell committed:
move legacy landsat 8 logic into mtl_metadata class

pjhartzell committed:
add title to doi Link

pjhartzell committed:
add doi links to collection summaries field

pjhartzell committed:
change how scientific doi links are included in collections

pjhartzell committed:
remove sr_ and st_ prefixes from asset keys

pjhartzell committed:
remove "unitless" units from raster extension

pjhartzell committed:
fix: finish removing "unitless" unit from oli_tirs

pjhartzell committed:
remove eo extension center_wavelength and full_width_half_max in collection item_assets

pjhartzell committed:
fix a slew of errors and add spatial_resolution to raster extension

pjhartzell committed:
update gsd

pjhartzell committed:
fix: remove 'name' field from MSS raster extension

pjhartzell committed:
fix: update fragments with traceable info.

pjhartzell committed:
change eo:bands layout, other edits to meet requested format

pjhartzell committed:
correct swir16 asset description in TM and ETM fragments

pjhartzell committed:
sundry fragment edits

pjhartzell committed:
condense MSS eo:bands summary, move band designation to description

pjhartzell committed:
attempt to better standardize and explain eo and asset information

pjhartzell committed:
move back to repeated bands in eo:bands summary

pjhartzell committed:
fix doi colon error

pjhartzell committed:
update c2-l1 examples

pjhartzell committed:
add new stactools antimeridian check

pjhartzell committed:
add all extensions to collection fragments

pjhartzell committed:
add antimeridian_strategy to commands

pjhartzell committed:
update the example for the legacy comparison

gadomski commented on src/stactools/landsat/commands.py:
```diff
@@ -74,12 +82,17 @@ def create_item_cmd(level: str, mtl: str, output: str, usgs_geometry: bool,
             level (str): Choice of 'level-1' or 'level-2'. This is not used
                 and has no effect.
         """
+        antimeridian = Antimeridian[antimeridian_strategy.upper()]
```
Do you think we should move this strategy stuff up to `stactools`?

pjhartzell replied:
The enum class?

pjhartzell replied:
Not sure what you mean...

gadomski replied:
We could make stactools function something like this `stactools.core.utils.antimeridian.fix_item_geometry(item, strategy=AntimeridianStrategy` that would pick whether to normalize or split it on the item.

pjhartzell replied:
Yeah, that would be cleaner, for sure.

pjhartzell committed:
remove problematic docker job from CI

pjhartzell committed:
chore: clean up .gitignore

pjhartzell committed:
chore: change stactools dependency to point at commit hash rather than branch

pjhartzell committed:
chore: update docs

- update ipynb
- remove the legacy l8 item directory from examples directory
- update command function documentation

pjhartzell committed:
chore: update changelog

gadomski commented on .gitignore:
```diff
@@ -1,5 +1,3 @@
-tests/data-files/external
-
```
Why was this directory removed? Generally the external data files _shouldn't_ be checked in.

pjhartzell replied:
Ignorance. I thought I added it during some of my testing. Didn't remember that it was part of the template .gitignore. Will add back.

pjhartzell committed:
fix: add tests/data-files/external back to .gitignore

pjhartzell committed:
chore: clean up changelong
